src/parse.rs

```diff
@@ -307,7 +302,15 @@ impl<'a> Parser<'a> {
                 None
             };
             if let Some(group) = group {
-                return Ok((ix + skip, Expr::Backref(group)));
+                self.backrefs.insert(group);
+                return Ok((
+                    ix + skip,
+                    if is_subroutine_call {
+                        Expr::SubroutineCall(group)
+                    } else {
+                        Expr::Backref(group)
+                    },
```

Comment:
I'm not sure about this `boolean` `is_subroutine_call` argument though. Would this be nicer?:

```rust
    fn parse_named_backref<F>(
        &self,
        ix: usize,
        open: &str,
        close: &str,
        allow_relative: bool,
        create_expr: F,
    ) -> Result<(usize, Expr)> where F: FnOnce(usize) -> Expr {
        ...
    }
```

And then either:

```rust
return self.parse_named_backref(end, "'", "'", true, |group| Expr::Backref(group));
```

or:

```rust
self.parse_named_backref(end, "'", "'", true, |group| Expr::SubroutineCall(group))?
```

Reply:
thanks for the great suggestion! I agree, I much prefer this way of doing things, it does look nicer - updated :)